vncviewer/GestureHandler.cxx

```diff
@@ -0,0 +1,439 @@
+const unsigned GH_TAP_TIMEOUT = 1000;
+
+// Timeout when waiting for longpress (ms)
+const unsigned GH_LONGPRESS_TIMEOUT = 1000;
+
+GestureHandler::GestureHandler() :
+  state(GH_INITSTATE), waitingRelease(false), longpressTimer(this) {
+}
+
+GestureHandler::~GestureHandler()
+{
+}
+
+void GestureHandler::handleTouchBegin(int id, double x, double y)
+{
+  GHTouch ght;
+
+  // Ignore any new touches if there is already an active gesture,
+  // or we're in a cleanup state
+  if (hasDetectedGesture() || (state == GH_NOGESTURE)) {
+    ignored.insert(id);
+    return;
+  }
+
+  // Did it take too long between touches that we should no longer
+  // consider this a single gesture?
+  if ((tracked.size() > 0) &&
+      (rfb::msSince(&tracked.begin()->second.started) > GH_MULTITOUCH_TIMEOUT)) {
+    state = GH_NOGESTURE;
+    ignored.insert(id);
+    return;
+  }
+
+  // If we're waiting for fingers to release then we should no longer
+  // recognize new touches
+  if (waitingRelease) {
+    state = GH_NOGESTURE;
+    ignored.insert(id);
+    return;
+  }
+
+  gettimeofday(&ght.started, NULL);
+  ght.active = true;
+  ght.last_x = ght.first_x = x;
+  ght.last_y = ght.first_y = y;
+
+  tracked[id] = ght;
+
+  switch (tracked.size()) {
+    case 1:
+      longpressTimer.start(GH_LONGPRESS_TIMEOUT);
+      break;
+
+    case 2:
+      state &= ~(GH_ONETAP | GH_DRAG | GH_LONGPRESS);
+      longpressTimer.stop();
+      break;
+
+    case 3:
+      state &= ~(GH_TWOTAP | GH_TWODRAG | GH_PINCH);
+      break;
+
+    default:
+      state = GH_NOGESTURE;
+  }
+}
+
+void GestureHandler::handleTouchUpdate(int id, double x, double y)
+{
+  GHTouch *touch;
+
+  // If this is an update for a touch we're not tracking, ignore it
+  if (tracked.count(id) == 0)
+    return;
+
+  touch = &tracked[id];
+
+  // Update the touches last position with the event coordinates
+  touch->last_x = x;
+  touch->last_y = y;
+
+  if (!hasDetectedGesture()) {
+    // If the move is smaller than the minimum threshold, ignore it
+    if (fabs(touch->first_x - x) < GH_MTHRESHOLD &&
+        fabs(touch->first_y - y) < GH_MTHRESHOLD)
+      return;
+
+    // Can't be a tap or long press as we've seen movement
+    state &= ~(GH_ONETAP | GH_TWOTAP | GH_THREETAP | GH_LONGPRESS);
+    longpressTimer.stop();
+
+    if (tracked.size() != 1)
+      state &= ~(GH_DRAG);
+    if (tracked.size() != 2)
+      state &= ~(GH_TWODRAG | GH_PINCH);
+
+    // We need to figure out which of our different two touch gestures
+    // this might be
+    if (tracked.size() == 2) {
+      double h, v, fdx, fdy, ldx, ldy, fd, ld, dt;
+
+      // Calculate both types of movement and pick the one that has
+      // moved the furthest
+
+      getAverageMovement(&h, &v);
+      h = fabs(h);
+      v = fabs(v);
+
+      getAverageDistance(&fdx, &fdy, &ldx, &ldy);
+      fd = sqrt(fdx * fdx + fdy * fdy);
+      ld = sqrt(ldx * ldx + ldy * ldy);
+      dt = fabs(ld - fd);
```

Comment:
We don't get one event with an update for every touch point, even if two fingers move simultaneously we get these updates spread in two events. That leads to cases where we get false PINCH gestures.

One such case is a fast two-finger-drag. Let's imagine two touches with a starting-distance of 10 pixels in between. Positions of the touches could be (0,10) and (0,0). If this was followed by a very fast movement in the same direction of both the fingers. The expected behavior would be a fast scroll.

The touch update for the first finger could have the position (0,80), and the second finger would still have (0,0). That means the movement for the first touch is larger than the threshold. In that case we will have these values:

    fd = 10
    ld = 70
    dt = 60

The `getAverageMovement()` function will see a movement of 70 on the first finger, and a movement of 0 for the second, and thus return:

    h = 0
    v = 35

And as we can see, (35 < 60) && (0 < 60) that means we eliminate `GH_TWODRAG`.

That's one example, there are other cases that cause this to happen as well. The problem is that the code was written with an incorrect assumption that we can a complete update of all positions simultaneously.

Perhaps it's too late for a complete rework, we should look into implementing a workaround for this. Perhaps we should wait some time before we can decide between PINCH and TWODRAG? That way we could allow potential "second-finger update" events to happen.

Reply:
fixed in 4f644ac